include/psa/crypto_values.h

```diff
@@ -1667,6 +1699,56 @@
 #define PSA_ALG_TLS12_PSK_TO_MS_GET_HASH(hkdf_alg)                         \
     (PSA_ALG_CATEGORY_HASH | ((hkdf_alg) & PSA_ALG_HASH_MASK))
 
+/* This flag indicates whether the key derivation algorithm is suitable for
+ * use on low-entropy secrets such as password - these algorithms are also
+ * known as key stretching or password hashing schemes. These are also the
+ * algorithms that accepts inputs of type #PSA_KEY_DERIVATION_INPUT_PASSWORD.
+ */
+#define PSA_ALG_KEY_DERIVATION_STRETCHING_FLAG  ((psa_algorithm_t)0x00008000)
```

Comment:
I'm sorry, this is a bit beyond me. Could you explain again like I'm not familiar with the context?

Reply:
For better or for worse, an algorithm can encode both a key agreement and a key derivation, for example `PSA_ALG_KEY_AGREEMENT(PSA_ALG_ECDH, PSA_ALG_HKDF(PSA_ALG_SHA_256))`. To use such an algorithm:
* `psa_key_derivation_setup` with the algorithm
* `psa_key_derivation_key_agreement` to perform ECDH
* Further `psa_key_derivation_xxx` to perform HKDF where the secret input is the shared secret from the ECDH part.

A key stretching has its own way of inputting the secret which is not compatible with a key agreement. In the encoding of key derivation and key agreement algorithms, bits 0..15 encode the KDF and bits 16..23 encode the key agreement. A key stretching doesn't need to be compatible with a key agreement, so the encoding of key stretching algorithms can use bits 16..23. Key derivation algorithms that are meant to be usable for the output of a key agreement can't use bits 16..23 in their encoding.

Therefore I propose to use a bit from 16..23 to indicate “this is a key stretching KDF”. There's a separate bit in 24..31 that indicates “this is a key derivation without a key agreement”, so there's no ambiguity.

Reply:
It would be good to ensure that if an application incorrectly attempted to combine a key agreement with a stretching KDF, that the result is not a known/defined composite algorithm. At present, the encoding scheme has zero in bits 23 and 22 for a key agreement, and encodes the key agreement type in bits 21:16.

Thus using a value that sets one or both of bits 23 and 22 to one should have the desired property. I think my preference is to designate bit 23 (as in @gilles-peskine-arm 's suggestion) as the key-stretching-KDF indicator. (All of the currently specified KDFs expect high-entropy secrets and so match this addition to the encoding scheme).